apollo-router/src/protocols/websocket.rs

```diff
@@ -361,7 +388,129 @@ where
         })
 }
 
-impl<S> Stream for GraphqlWebSocket<S>
+pub(crate) struct SubscriptionStream<S> {
+    inner_stream: SplitStream<InnerStream<S>>,
+    close_signal: Option<tokio::sync::oneshot::Sender<()>>,
+}
+
+impl<S> SubscriptionStream<S>
+where
+    S: Stream<Item = serde_json::Result<ServerMessage>>
+        + Sink<ClientMessage>
+        + std::marker::Unpin
+        + std::marker::Send
+        + 'static,
+{
+    pub(crate) fn new(
+        stream: S,
+        id: String,
+        protocol: WebSocketProtocol,
+        heartbeat_interval: Option<tokio::time::Duration>,
+    ) -> Self {
+        let (mut sink, inner_stream) = InnerStream::new(stream, id, protocol).split();
+        let (close_signal, close_sentinel) = tokio::sync::oneshot::channel::<()>();
+
+        tokio::task::spawn(async move {
+            if let (WebSocketProtocol::GraphqlWs, Some(duration)) = (protocol, heartbeat_interval) {
+                let mut interval =
+                    tokio::time::interval_at(tokio::time::Instant::now() + duration, duration);
+                interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+                let mut heartbeat_stream = IntervalStream::new(interval)
+                    .map(|_| {
+                        Ok(ClientMessage::Ping {
+                            payload: Some(serde_json_bytes::Value::String(
+                                "APOLLO_ROUTER_HEARTBEAT".into(),
+                            )),
+                        })
+                    })
+                    .take_until(close_sentinel);
+                if let Err(err) = sink.send_all(&mut heartbeat_stream).await {
+                    tracing::trace!("cannot send heartbeat: {err:?}");
+                    if let Some(close_sentinel) = heartbeat_stream.take_future() {
+                        if let Err(err) = close_sentinel.await {
+                            tracing::trace!("cannot shutdown sink: {err:?}");
+                        }
+                    }
+                }
+            } else if let Err(err) = close_sentinel.await {
+                tracing::trace!("cannot shutdown sink: {err:?}");
+            };
+
+            tracing::info!(
+                monotonic_counter
+                    .apollo
+                    .router
+                    .operations
+                    .subscriptions
+                    .events = 1u64,
+                subscriptions.mode = "passthrough",
+                subscriptions.complete = true
+            );
+
+            if let Err(err) = sink.close().await {
+                tracing::trace!("cannot close the websocket stream: {err:?}");
+            }
+        });
+
+        Self {
+            inner_stream,
+            close_signal: Some(close_signal),
+        }
+    }
+}
+
+impl<S> Drop for SubscriptionStream<S> {
+    fn drop(&mut self) {
+        if let Some(close_signal) = self.close_signal.take() {
```

Comment:
technically you don't even need to send with `close_signal`. If it is dropped, `close_sentinel` will resolve and return the error `TryRecvError::Closed`: https://docs.rs/tokio/latest/tokio/sync/oneshot/error/enum.TryRecvError.html
but it is good to be explicit